agent/connect/x509_patch.go

```diff
@@ -0,0 +1,122 @@
+package connect
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"fmt"
+	"net"
+	"net/url"
+	"unicode"
+)
+
+// NOTE: the contents of this file were lifted from
+// $GOROOT/src/crypto/x509/x509.go from a Go 1.16.5 checkout.
+//
+//
+// After https://go-review.googlesource.com/c/go/+/329129 lands in a Go release
+// we are compiling against we can safely remove all of this code.
+
+var (
+	x509_oidExtensionSubjectAltName = []int{2, 5, 29, 17}
+)
+
+const (
+	x509_nameTypeEmail = 1
+	x509_nameTypeDNS   = 2
+	x509_nameTypeURI   = 6
+	x509_nameTypeIP    = 7
+)
+
+func HackSANExtensionForCSR(template *x509.CertificateRequest) {
+	switch {
+	case len(template.DNSNames) > 0:
+	case len(template.EmailAddresses) > 0:
+	case len(template.IPAddresses) > 0:
+	case len(template.URIs) > 0:
+	default:
+		return
+	}
+
+	if x509_oidInExtensions(x509_oidExtensionSubjectAltName, template.ExtraExtensions) {
+		return
+	}
+
+	// Ensure we wipe the subject field completely since we're going to use the
+	// SAN field instead.
+	template.RawSubject = nil
+	template.Subject = pkix.Name{}
+
+	value, err := x509_marshalSANs(template.DNSNames, template.EmailAddresses, template.IPAddresses, template.URIs)
+	if err != nil {
+		return
+	}
+
+	ext := pkix.Extension{
+		Id: x509_oidExtensionSubjectAltName,
+		// From RFC 5280, Section 4.2.1.6:
+		// “If the subject field contains an empty sequence ... then
+		// subjectAltName extension ... is marked as critical”
+		//
+		// Since we just cleared the subject above, it's critical.
+		Critical: true,
+		Value:    value,
+	}
+	template.ExtraExtensions = append(template.ExtraExtensions, ext)
+}
+
+// oidNotInExtensions reports whether an extension with the given oid exists in
+// extensions.
+func x509_oidInExtensions(oid asn1.ObjectIdentifier, extensions []pkix.Extension) bool {
+	for _, e := range extensions {
+		if e.Id.Equal(oid) {
+			return true
+		}
+	}
+	return false
+}
+
+// marshalSANs marshals a list of addresses into a the contents of an X.509
```

Comment:
Small nit: `marshalSANs` is not the real name of the function anymore.

```suggestion
// x509_marshalSANs marshals a list of addresses into a the contents of an X.509
```